blocks/search.py

```diff
@@ -0,0 +1,280 @@
+        with names as returned by `generate_outputs`.
+
+        """
+        # First timesteps only if state is needed
+        states = [cur_states[name][0] for name in self.need_input_states]
+
+        next_values = self.next_state_computer(*(list(contexts.values()) +
+                                                 states))
+
+        # Add time dimension back
+        next_values = [state.reshape((1,) + state.shape)
+                       for state in next_values]
+        return OrderedDict(zip(self.generate_names + ['probs'], next_values))
+
+    @staticmethod
+    def _top_probs(probs, beam_size):
+        """Returns indexes of elements with highest probabilities.
+
+        Parameters
+        ----------
+        probs : numpy array
+            A 3d array of probabilities (length of sequence, batch,
+            readout_dim).
+        beam_size : int
+            Beam size, number of top probabilities to return.
+
+        Returns
+        -------
+        Tuple of (indexes, top probabilities).
+
+        """
+        flatten = probs.flatten()
+        args = numpy.argpartition(-flatten, beam_size)[:beam_size]
+        args = args[numpy.argsort(-flatten[args])]
+        # convert args back
+        indexes = numpy.unravel_index(args, probs.shape[1:])
+        return indexes, probs[0][indexes]
+
+    @staticmethod
+    def _rearrange(outputs, indexes):
+        new_outputs = outputs[:, indexes.T.flatten()]
+        return new_outputs.copy()
+
+    def search(self, inputs_val_dict, eol_symbol=-1, max_length=512):
+        """Performs beam search.
+
+        If the beam search was not compiled, it also compiles it.
+
+        Parameters
+        ----------
+        inputs_val_dict : dict
+            Dictionary of input values {name: value}. Batch size `1` is
+            supported only.
+        eol_symbol : int
+            End of sequence symbol, the search stops when the
+            symbol is generated.
+        max_length : int
+            Maximum sequence length, the search stops when it
+            is reached.
+
+        Returns
+        -------
+        Sequences in the beam, masks, and corresponding probabilities.
+
+        """
+        if not self.compiled:
+            self.compile()
+        # Inputs repeated beam_size times
+        inputs = OrderedDict([(name, numpy.tile(val, (1, self.beam_size)))
+                              for name, val in inputs_val_dict.items()])
+
+        # Precompute contexts
+        contexts = self.compute_contexts(inputs)
+
+        states = self.compute_initial_states(contexts)
+
+        states['cur_outputs'] = states['outputs']
+        states['cur_outputs_mask'] = numpy.ones_like(states['cur_outputs'])
+        states['cur_probs'] = numpy.ones_like(states['cur_outputs'])
+
+        for i in range(max_length):
+            states.update(self.compute_next(contexts, states))
+            next_probs = (states['cur_probs'][:, :, None] * states['probs'] **
```

Comment:
Okay, the reason for different scores is clear: you are working with pure probabilities here. Please use logarithms, this much safer.